### PR TITLE
Persistence Layer and Module binding

### PR DESCRIPTION
**Pull request by dlgn2**

### CODE REVIEW

File: src/implementations/CrunchDB.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { Container } from 'inversify';
+import { ICrunchDB } from 'crunchDB/interfaces/ICrunchDB';
```
Comment (sdmuki) [inline src/implementations/CrunchDB.ts]: import { ICrunchDB } from 'crunchDB/interfaces';
# use module imports from sibling directories

File: src/implementations/business/volatile/IPersistenceConfigProvider.ts [hunk 2]
```diff
@@ -0,0 +1,11 @@
+import { ResultAsync } from 'neverthrow';
+
+import { IPersistenceConfig } from 'crunchDB/implementations/business/volatile/IPersistenceConfig.js';
```
Comment (sdmuki) [inline src/implementations/business/volatile/IPersistenceConfigProvider.ts]: no js imports.

File: src/implementations/business/volatile/IVolatileStorage.ts [hunk 3]
```diff
@@ -0,0 +1,117 @@
+import {
+  PersistenceError,
+  VersionedObject,
+  VolatileStorageMetadata,
+  VolatileStorageKey,
+  ERecordKey,
+} from 'crunchDB/objects';
+import { ResultAsync } from 'neverthrow';
+
+import { IVolatileCursor } from 'crunchDB/implementations/business/volatile/IVolatileCursor.js';
```
Comment (sdmuki) [inline src/implementations/business/volatile/IVolatileStorage.ts]: no js imports

File: src/implementations/business/volatile/IndexedDB.ts [hunk 4]
```diff
@@ -0,0 +1,1239 @@
...
+  VersionedObject,
+  VolatileStorageDataKey,
+  VolatileStorageKey,
+  VolatileStorageMetadata,
+  VolatileStorageMetadataIndexes,
+} from 'crunchDB/objects';
+import { errAsync, okAsync, ResultAsync } from 'neverthrow';
+import { ResultUtils } from 'neverthrow-result-utils';
+
+import { IndexedDBCursor } from 'crunchDB/implementations/business/volatile/IndexedDBCursor.js';
```
Comment (sdmuki) [inline src/implementations/business/volatile/IndexedDB.ts]: no js imports

File: src/implementations/business/volatile/VolatileTableIndex.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
+import {
+  EBackupPriority,
+  ERecordKey,
+  VersionedObject,
+  VersionedObjectMigrator,
+} from 'crunchDB/objects';
+
+import { IStorageIndex } from 'crunchDB/implementations/business/volatile/IStorageIndex.js';
```
Comment (sdmuki) [inline src/implementations/business/volatile/VolatileTableIndex.ts]: js imports

File: src/objects/business/VolatileDataUpdate.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
+import { VersionedObject } from "@objects/businessObjects/versioned/index.js";
```
Comment (sdmuki) [inline src/objects/business/VolatileDataUpdate.ts]: js imports not needed. use "objects". get rid of the @. We will use @ for domains only which is not decided yet

File: src/objects/business/VolatileDataUpdate.ts [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+import { EDataUpdateOpCode } from "@objects/enum/index.js";
```
Comment (sdmuki) [inline src/objects/business/VolatileDataUpdate.ts]: objects/enum

File: src/objects/business/VolatileDataUpdate.ts [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+import {
+  UnixTimestamp,
+  VolatileStorageKey,
+} from "@objects/primitives/index.js";
```
Comment (sdmuki) [inline src/objects/business/VolatileDataUpdate.ts]: objects/primitives

**Review by sdmuki — CHANGES_REQUESTED** [review]

File: package-lock.json
Comment (SnickerChar) [inline package-lock.json:1]: package-lock.json implies using NPM, you should probably be using Yarn

File: src/interfaces/utilities/ILogUtils.ts [hunk 9]
```diff
@@ -0,0 +1,13 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
+import pino from "pino";
+
+export interface ILogUtils {
+  debug(message?: any, ...optionalParams: any[]): void;
+  info(message?: any, ...optionalParams: any[]): void;
+  log(message?: any, ...optionalParams: any[]): void;
+  warning(message?: any, ...optionalParams: any[]): void;
+  error(message?: any, ...optionalParams: any[]): void;
+  getPino(): pino.Logger;
```
Comment (SnickerChar) [inline src/interfaces/utilities/ILogUtils.ts:10]: You can remove this method for sure. It's been a carry-forward for way too long.

File: src/implementations/business/volatile/VolatileStorageSchemaProvider.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
+import {
```
Comment (SnickerChar) [inline src/implementations/business/volatile/VolatileStorageSchemaProvider.ts:1]: The storage schema for the DB needs to be provided via the CrunchDB constructor; it's not something that will be configured or known before you initialize the CrunchDB

File: src/objects/business/VersionedObject.ts [hunk 11]
```diff
@@ -0,0 +1,27 @@
+import { PropertiesOf } from 'crunchDB/objects/utilities';
```
Comment (SnickerChar) [inline src/objects/business/VersionedObject.ts:1]: VersionedObject is not something that CrunchDB should worry about. It should be provided with the current schema on initialization.

File: src/objects/business/VolatileDataUpdate.ts [hunk 12]
```diff
@@ -0,0 +1,17 @@
+import {
+  EDataUpdateOpCode,
+  VolatileStorageKey,
+  UnixTimestamp,
+  VersionedObject,
+} from 'crunchDB/objects';
+
```
Comment (SnickerChar) [inline src/objects/business/VolatileDataUpdate.ts:7]: CrunchDB shouldn't be updating the DB

File: src/objects/enum/EBackupPriority.ts [hunk 13]
```diff
@@ -0,0 +1,5 @@
+export enum EBackupPriority {
```
Comment (SnickerChar) [inline src/objects/enum/EBackupPriority.ts:1]: All the things related to backups should be nuked.

**Review by SnickerChar — COMMENTED** [review]
There's a lot of stuff you are bringing in that is irrelevant to CrunchDB. It's not going to be doing backups or versioning, for instance.

**Pull request closed without merging**